source: implement FeatureSource interface

marquiz opened the pull request:
Convert the `cpu`, `kernel`, `pci`, `usb`, `system`, `custom` and `local` sources to implement the `FeatureSource` interface, i.e. do feature discovery and creation of feature labels separately.

Changes the related custom rules to utilize `GetFeatures()` method of the feature sources.

Also moves different "utils" from `source/internal` to the respective sources.

This PR is split from #464, to make review easier. It demonstrates usage of the new `FeatureSource` interface with some of the sources. Other sources will be converted in future PRs.

marquiz commented:
Rebased. Depends on #642 
/hold

marquiz commented:
/unhold

marquiz commented:
/assign @zvonkok @ArangoGutierrez

zvonkok commented:
@marquiz This fails for me: 

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xc3c369]

goroutine 1 [running]:
sigs.k8s.io/node-feature-discovery/source/kernel.(*kernelSource).GetLabels(0x157b220)
	/go/node-feature-discovery/source/kernel/kernel.go:108 +0x269
sigs.k8s.io/node-feature-discovery/pkg/nfd-client/worker.getFeatureLabels({0x7ff284548b88, 0x157b220}, {{0x0, 0x0}, 0xc0002e0c60, 0x0, 0x0, 0x15555, {0xc000276fa0, 0x1, ...}, ...})
	/go/node-feature-discovery/pkg/nfd-client/worker/nfd-worker.go:427 +0xb0
sigs.k8s.io/node-feature-discovery/pkg/nfd-client/worker.createFeatureLabels({0xc0001e73f0, 0xb, 0x203000}, {{0x0, 0x0}, 0xc0002e0c60, 0x0, 0x0, 0x15555, {0xc000276fa0, ...}, ...})
	/go/node-feature-discovery/pkg/nfd-client/worker/nfd-worker.go:408 +0x1d9
sigs.k8s.io/node-feature-discovery/pkg/nfd-client/worker.(*nfdWorker).Run(0xc0001a5ce0)
	/go/node-feature-discovery/pkg/nfd-client/worker/nfd-worker.go:187 +0xbb8
main.main()
	/go/node-feature-discovery/cmd/nfd-worker/main.go:63 +0x371
````

zvonkok commented:
I think this comes from line 139 in kernel.go. When SELinux is not enabled there is no entry added.

marquiz commented:
> I think this comes from line 139 in kernel.go. When SELinux is not enabled there is no entry added.

Argh, my latest change (#642) causes these failures (selinux is not the only possible point of panic 😬 ). I'll fix and add unit tests for this. Thanks for testing @zvonkok!

marquiz commented:
OK, this should be fixed, now. Fixed by reverting part of my earlier internal api change.

I also added simple unit tests to the feature sources to avoid this.

ArangoGutierrez commented:
e2e-test works on Minikube 1.22
```bash
Ran 2 of 230 Specs in 24.488 seconds
SUCCESS! -- 2 Passed | 0 Failed | 0 Pending | 228 Skipped
--- PASS: TestE2E (24.51s)
PASS
ok      sigs.k8s.io/node-feature-discovery/test/e2e     24.543s
```
test
```bash
go test ./cmd/... ./pkg/... ./source/...                                                                                                                                                                                                      
?       sigs.k8s.io/node-feature-discovery/cmd/nfd-master       [no test files]                                                                                                                                                               
ok      sigs.k8s.io/node-feature-discovery/cmd/nfd-topology-updater     (cached)                          
ok      sigs.k8s.io/node-feature-discovery/cmd/nfd-worker       0.012s                                                 
?       sigs.k8s.io/node-feature-discovery/pkg/api/feature      [no test files]                     
?       sigs.k8s.io/node-feature-discovery/pkg/apihelper        [no test files]                                        
?       sigs.k8s.io/node-feature-discovery/pkg/cpuid    [no test files]        
ok      sigs.k8s.io/node-feature-discovery/pkg/kubeconf (cached)                                                       
?       sigs.k8s.io/node-feature-discovery/pkg/labeler  [no test files]                                                
?       sigs.k8s.io/node-feature-discovery/pkg/nfd-client       [no test files]                                        
ok      sigs.k8s.io/node-feature-discovery/pkg/nfd-client/topology-updater      (cached)                                                                                                                                                      
ok      sigs.k8s.io/node-feature-discovery/pkg/nfd-client/worker        3.093s                                         
ok      sigs.k8s.io/node-feature-discovery/pkg/nfd-master       (cached)                                               
?       sigs.k8s.io/node-feature-discovery/pkg/podres   [no test files]                                                
ok      sigs.k8s.io/node-feature-discovery/pkg/resourcemonitor  (cached)                                          
?       sigs.k8s.io/node-feature-discovery/pkg/topologypolicy   [no test files]                                        
?       sigs.k8s.io/node-feature-discovery/pkg/topologyupdater  [no test files]                                        
ok      sigs.k8s.io/node-feature-discovery/pkg/utils    (cached)                                                       
?       sigs.k8s.io/node-feature-discovery/pkg/version  [no test files]                                                                                                                                                                       
ok      sigs.k8s.io/node-feature-discovery/source       0.033s                                                         
ok      sigs.k8s.io/node-feature-discovery/source/cpu   0.034s                                                         
?       sigs.k8s.io/node-feature-discovery/source/custom        [no test files]                                        
?       sigs.k8s.io/node-feature-discovery/source/custom/rules  [no test files]                                                                                                                                                               
?       sigs.k8s.io/node-feature-discovery/source/fake  [no test files]                                                                                                                                                                       ?       sigs.k8s.io/node-feature-discovery/source/iommu [no test files]                                                
ok      sigs.k8s.io/node-feature-discovery/source/kernel        0.038s                                                 
ok      sigs.k8s.io/node-feature-discovery/source/local 0.030s                                                         
?       sigs.k8s.io/node-feature-discovery/source/memory        [no test files]                                        
?       sigs.k8s.io/node-feature-discovery/source/network       [no test files]                         
ok      sigs.k8s.io/node-feature-discovery/source/pci   0.007s                                          
?       sigs.k8s.io/node-feature-discovery/source/storage       [no test files]                                        
ok      sigs.k8s.io/node-feature-discovery/source/system        0.004s                                                 
ok      sigs.k8s.io/node-feature-discovery/source/usb   0.004s   
```

ArangoGutierrez reviewed (approved):
/lgtm
